rllib/algorithms/algorithm.py

```diff
@@ -3209,6 +3204,69 @@ def _automatic_evaluation_duration_fn(
                 * eval_cfg["num_envs_per_worker"]
             )
 
+    @staticmethod
+    def _evaluate_async_remote_fn(
+        _worker: EnvRunner,
+        _weights_ref: ray.ObjectRef,
+        _weights_seq_no: int,
+        _env_runner: bool = False,
```

Comment:
Do we need to pass in these arguments. `_worker` gives already this information to the function body.

Reply:
Yes, we "bake" them in via a partial, which we then send to the ActorManager.
The `_worker` arg is provided "automatically" by the `ActorManager.foreach_worker_async` utility. So we don't have to bake that into the partial. Is that what you meant?

By removing the "closure" (local `weights_ref` var inside the function body pointing to a `ray.put()`-returned ray ObjectRef outside(!) the function) from the function body, we fix the memory leak.

Reply:
Yeah this I got. `weight_ref` and `weight_seq_no` are necessary to avoid later the memory leak. I meant the `env_runner` and `env_runner_num_episodes`, but specifically the first of the two. I mean `issubclass(workers, EnvRunner)` could give us the same inside of the function, couldn't it?

Reply:
oh, sorry, this is just a flag indicating that we need to sample in a certain way. There is a small difference whether you call `sample()` on RolloutWorker (returns batch) or on Single/MultiAgentEnvRunner (return lists of episodes).
Also, for evaluation, if we count by episodes, one evaluation round should always only sample a single episode.